File: arrays-and-strings/one_away.py
# 1.5: One Away

# Runtime: O(n) - Space: O(1) * technically O(n) because of longer/shorter
def one_away(s1: str, s2: str) -> bool:
    if abs(len(s1) - len(s2)) > 1:
        return False

    longer = s1 if len(s1) >= len(s2) else s2
    shorter = s1 if len(s1) < len(s2) else s2
    differences = 0

    for s in shorter:
        if s not in longer:
            differences += 1

    return differences <= 1


print(one_away("pale", "ple"))  # True
print(one_away("pales", "pale"))  # True
print(one_away("pale", "bale"))  # True
print(one_away("pale", "bake"))  # False

File: 01-arrays-and-strings/javascript/oneAway.js
// 1.5: One Away

// Runtime: O(min(n, m)) - Space: O(1)
function oneAway(s1, s2) {
  if (Math.abs(s1.length - s2.length) > 1) return false

  const longer = s1.length >= s2.length ? s1 : s2
  const shorter = s1.length < s2.length ? s1 : s2
  let differences = 0

  for (const char in shorter) {
    if (!s2.match(longer[char])) differences += 1
  }

  return differences <= 1
}

console.log(oneAway('pale', 'ple'))  // true
console.log(oneAway('pales', 'pale'))  // true
console.log(oneAway('pale', 'bale'))  // true
console.log(oneAway('pale', 'bake'))  // false
